Fixes #845: NPCs hold torches during the day

lgromanowski opened the pull request:
Added method in WeatherManger and World which returns true if it is night.
This method is used later in character controller to show torches
(or other sources of light) at night and hide them at day.

Signed-off-by: Lukasz Gromanowski lgromanowski@gmail.com

lgromanowski committed:
Fixes #845: NPCs hold torches during the day

Added method in WeatherManger and World which returns true if it is night.
This method is used later in character controller to show torches
(or other sources of light) at night and hide them at day.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

kcat commented:
A better idea would be to unequip torches, rather than forcing them to not show despite being equipped.

lgromanowski commented:
Ok, I will try to do that with equiping / unequiping.

lgromanowski committed:
Fixes #845: NPCs hold torches during the day

Simplified a bit code which shows and hides light.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

lgromanowski commented:
I tried with equipping/unequipping but it doesn't work as expected (after unequipping NPC couldn't equip light again). This is the best result I found - if NPC have already equipped light, then play the animation, and then hide them.

kcat commented:
Why couldn't they equip the light again? If the actor AI properly unequips lights while the NPC is outside during the day, it should be able to equip again at night. This is still a very poor way to do it since it's just hiding the fact that the actor has a torch, so anything else that checks if something is equipped in the left hand will still see it.

lgromanowski committed:
Fixes #845: NPCs hold torches during the day

Added equipping/unequipping torches.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

lgromanowski commented:
Ok, it is working now (I was searching inventory for wrong torch after unequipping, that's why it wasn't work).

kcat commented:
It would be better to do it in the AI routines. The character controller is just where it handles the animation states based on the current state of the actor, it shouldn't be modifying inventory and such. Also, you want to make sure the player isn't affected by this as he should be able to keep a torch equipped whenever he wants.

lgromanowski commented:
Player shouldn't be affected because it's in updateNpcState() method

kcat commented:
updateNpcState affects the player, too. The name is a bit of a misnomer in that regard, but it follows the naming scheme in both the game data and elsewhere in the code (e.g. the Npc mwclass and NpcAnimation). The player is technically an NPC actor type, instead of a creature actor type.

lgromanowski committed:
Fixes #845: NPCs hold torches during the day

Added check for Player character so it won't be affected by showing,
or hidding torches.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

lgromanowski commented:
You're right - I've just added check for player character.

zinnschlag commented:
This crashes for me upon entering Seyda Neen.

lgromanowski commented:
I tried to reproduce your crash by starting new game, going from / to seyda neen, but it didn't crashed. Could you give me more information what did you do to get the crash (and if you can reproduce it could you give me a backtrace)?

zinnschlag commented:
Can't provide a backtrace right now, because some of the recent changes make it impossible for me to run OpenMW under gdb. I PM'ed scrawl about it.

Steps to reproduce (hopeuflly):
- start OpenMW with --skip-menu
- either use --start "Seyda Neen" or coc "Seyda Neen". Both leads to an instant crash.

ghost commented:
1. MWMechanics::Actors would be a better place than CharacterController.
2. If there is no torch, the current implementation does a string search through the whole inventory of the actor every frame, which should be avoided. Better remember if a torch is present in the inventory or not. You could add a flag for that (or a ContainerStoreIterator to the torch) to InventoryStore and update it whenever the inventory is modified.
3. Now that it's equipped/unequipped instead of shown/hidden, showLights needs to be removed.
4. for consistency, autoEquip should be modified to no longer equip lights.

zinnschlag commented:
Okay, that problem is not in your branch. It is in master. We need to figure that out first, before we can do any merging.
For reference, here is the backtrace: https://gist.github.com/zinnschlag/8060447

ghost commented:
You are using nouveau. Is that intended?
And just out of curiosity, what does the assertion say (assume it's an assertion)?

zinnschlag commented:
Yeah, I am using nouveau.

No assertion. I am getting: 

Program received signal SIGABRT, Aborted.
0x00007ffff3d2d445 in __GI_raise (sig=<optimized out>) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
64  ../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.

ghost commented:
That's weird. Would you be able to narrow it down to a commit? I don't think anything changed recently in our graphics code. Perhaps a driver update caused it?

coc "seyda neen" works for me.

zinnschlag commented:
I did not have any driver update recently. Will try to identify the commit, but that might take a while. From Monday on a will be away from Christmas and I am not sure if I can get to it this weekend.

zinnschlag commented:
However my savedgame branch does not show any crashes and I merged master into that quite recently.

ghost commented:
Then bisecting should hopefully not take long ;)

lgromanowski committed:
Fixes #845: NPCs hold torches during the day

Moved 'equipping torches at night and unequipping at day' code from Character
to Actors class.

Removed unneeded showLights method (introduced in previous commits)
from animation/npcanimation classes.

Since this commit autoEquip() method doesn't automatically equip lights.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

zinnschlag commented:
Finished bisecting. The culprit is:

> commit 6a3eb3b355ecff7983ad61eeaad01fb576568e3f
> Author: scrawl scrawl@baseoftrash.de
> Date:   Thu Dec 19 01:31:12 2013 +0100
> 
>    Better way of reversing layer UV.

ghost commented:
I am clueless. That line is perfectly valid. Seems like a driver bug. What mesa version do you use? Perhaps it's fixed in a later version.

As a workaround, you can disable the terrain shader using

[Terrain]
shader = false

in ~/.config/openmw/settings.cfg.

zinnschlag commented:
Mesa: 8.0.4-0ubuntu0.6. Up to date.

btw. sorry for hijacking this pull request with this problem.

ghost commented:
Just a note for next time, some files in this pull request have wrong indentation (2 spaces instead of 4). I'll just fix this since I wanted to make a few additions anyway.

lgromanowski commented on apps/openmw/mwmechanics/character.cpp:
```diff
@@ -36,6 +36,8 @@
 #include "../mwworld/class.hpp"
 #include "../mwworld/inventorystore.hpp"
+#include "../mwworld/actionequip.hpp"
+#include "../mwworld/actiontake.hpp"
```
Ouch, you're right, it should be removed.

lgromanowski commented on apps/openmw/mwworld/weather.cpp:
```diff
@@ -707,3 +707,8 @@ float WeatherManager::getWindSpeed() const
+
+bool WeatherManager::isNight() const
+{
+  return (mHour < mSunriseTime || mHour > mNightStart - 1);
```
I don't mind to change it to isDark() (I'll push a patch later).
